ethicalgardeners/renderer.py

```diff
@@ -0,0 +1,700 @@
+                )
+
+            self._run_id = int(time.time())
+
+        self._grid_world = None
+        self._agents = None
+
+    def render(self, grid_world, agents):
+        """
+        Render the current state of the grid world as text.
+
+        This method create a text representation of the grid world, using the
+        character mappings to display different elements. Each cell is
+        represented by a single character followed by numbers for plants and
+        agents, with rows separated by newlines. If post_analysis_on is True,
+        it will also create a graphical frame for later video export.
+
+        Args:
+            grid_world (:py:class:`.WorldGrid`): The current state of the world
+                grid to render.
+            agents (dict): Dictionary mapping the agent's gymnasium ID to
+                the Agent instance.
+        """
+        # Store references to grid_world and agents
+        self._grid_world = grid_world
+        self._agents = agents
+
+        # Create a grid representation of the world
+        self.grid_representation = []
+        for i in range(grid_world.height):
+            row = []
+            row.append("|")  # Start of row
+            for j in range(grid_world.width):
+                cell = grid_world.get_cell((i, j))
+
+                # Empty cell by default
+                cell_char = self.characters.get('ground', ' ')
+
+                # Check cell type and update character accordingly
+                if cell.cell_type == CellType.OBSTACLE:
+                    cell_char = self.characters.get('obstacle', 'O')
+
+                # Verify if the cell contains a flower
+                if cell.has_flower():
+                    cell_char = self.characters.get('flower', 'F')
+                    flower_type = cell.flower.flower_type
+                    growth_stage = cell.flower.current_growth_stage
+                    cell_char = f"{cell_char}{flower_type}_{growth_stage}"
+
+                # Verify if the cell contains an agent (above all)
+                if cell.has_agent():
+                    cell_char = self.characters.get('agent', 'A')
+                    agent_id = grid_world.agents.index(cell.agent)
+                    cell_char = f"{cell_char}{agent_id}"
+
+                # Add the pollution level
+                if cell.pollution is not None:
+                    cell_char += f" {cell.pollution}"
+                else:
+                    cell_char += "   "
+
+                row.append(cell_char)
+                row.append('|')  # Separator for cells
+            self.grid_representation.append(''.join(row))
+
+        # if post_analysis_on is True, we will create a frame for video export
+        if self.post_analysis_on:
+            try:
+                import pygame
+            except ImportError:
+                raise ImportError(
+                    "Error while importing pygame. "
```

Comment:
Same remark as for WandB: we should tell the user what to do. This message does not provide any information; there was an error, OK, what should I do next?

"PyGame is required to produce a video. Please install it with `pip install pygame` or `pip install ethicalgardeners[visu]`"

Additionally, I'm thinking that maybe we should not raise an error, but rather a warning and then exit the method? So that we avoid crashing the whole program. It's not that pygame is strictly required for running the experiment itself, but rather than we need it to generate the video. So, let's just not generate the video (because we can't), and move on.

This would avoid the maddening situation where we wait for hours for the simulation to finish, only to get an error at the end.